### DISCUSSION TITLE
cryptography

### DISCUSSION THREAD

**Discussion by khromenokroman** [strongswan/strongswan#1854] [General]
Guys, help me solve the question. I want to integrate your product into my solution, but I need to understand where you are working with encryption and decryption, as I skimmed through, it seems to me that working with cryptography goes through the operating system kernel, did I understand correctly?

**Comment by strongX509** [comment]
The IPsec ESP encryption/description of IP packets (IPsec SA) is done by the Linux kernel whereas the encryption/description of the IKEv2 tunnel setup negotiation (IKE_SA) is done by the strongSwan daemon.

**Reply by khromenokroman** [reply]
Thank you very much, you helped me a lot :)

**Reply by khromenokroman** [reply]
There is another question, is it possible to pull out the entire ControlPlane (IKE + SA) in UserSpace?

**Reply by khromenokroman** [reply]
I can't figure out yet how to decrypt data in UserSpace using IKE in the kernel

**Comment by strongX509** [comment]
The strongSwan IKE daemon (charon) is running in UserSpace. Only the IPsec SAs are installed in the Linux kernel via the XFRM kernel interface. If you want IPsec ESP UserSpace encryption/decryption then you can use the **`kernel-libipsec`** strongSwan plugin.

**Reply by khromenokroman** [reply]
Thank you very much, you helped me a lot :)
